util/container/Dockerfile

```diff
@@ -4,174 +4,127 @@
-RUN cargo install --path /tmp/banshee
+RUN git clone https://github.com/pulp-platform/banshee.git /tmp/banshee --recurse-submodules && \
+    cargo install --path /tmp/banshee
 
 # Install Bender
-RUN wget https://github.com/pulp-platform/bender/releases/download/v${BENDER_VERSION}/bender-${BENDER_VERSION}-x86_64-linux-gnu-ubuntu18.04.tar.gz
-RUN tar xzf bender-${BENDER_VERSION}-x86_64-linux-gnu-ubuntu18.04.tar.gz
+RUN wget https://github.com/pulp-platform/bender/releases/download/v${BENDER_VERSION}/bender-${BENDER_VERSION}-x86_64-linux-gnu-ubuntu${UBUNTU_VERSION}.tar.gz && \
+    tar xzf bender-${BENDER_VERSION}-x86_64-linux-gnu-ubuntu${UBUNTU_VERSION}.tar.gz
 
-# Install spike-dasm
-RUN wget https://github.com/pulp-platform/riscv-isa-sim/releases/download/snitch-v${SPIKE_DASM_VERSION}/snitch-spike-dasm-${SPIKE_DASM_VERSION}-x86_64-linux-gnu-ubuntu18.04.tar.gz
-RUN tar xzf snitch-spike-dasm-${SPIKE_DASM_VERSION}-x86_64-linux-gnu-ubuntu18.04.tar.gz
+# Install Verible
+RUN wget https://github.com/chipsalliance/verible/releases/download/v${VERIBLE_VERSION}/verible-v${VERIBLE_VERSION}-linux-static-x86_64.tar.gz && \
+    tar -x -f verible-v${VERIBLE_VERSION}-linux-static-x86_64.tar.gz --strip-components=1 -C . && \
+    rm -rf verible-v${VERIBLE_VERSION}-linux-static-x86_64.tar.gz
 
 # Install Doxygen
-RUN wget https://www.doxygen.nl/files/doxygen-${DOXYGEN_VERSION}.linux.bin.tar.gz
-RUN tar xzf doxygen-${DOXYGEN_VERSION}.linux.bin.tar.gz
-RUN mv doxygen-${DOXYGEN_VERSION} doxygen
+RUN wget https://www.doxygen.nl/files/doxygen-${DOXYGEN_VERSION}.linux.bin.tar.gz && \
+    tar xzf doxygen-${DOXYGEN_VERSION}.linux.bin.tar.gz && \
+    mv doxygen-${DOXYGEN_VERSION} doxygen
 
-# 2. Stage
-FROM ubuntu:22.04 AS snitch_cluster
-ARG SNITCH_LLVM_VERSION=latest
-ARG VERIBLE_VERSION=0.0-3318-g8d254167
+# Get the precompiled LLVM toolchain
+RUN latest_tag=`curl -s -H "Accept: application/vnd.github.v3+json" https://api.github.com/repos/pulp-platform/llvm-project/releases/latest | grep '"tag_name":' | sed -E 's/.*"([^"]+)".*/\1/'` && \
+echo "SNITCH_LLVM_VERSION=${SNITCH_LLVM_VERSION} LLVM_TAR=${LLVM_TAR} latest_tag=${latest_tag}" && \
+test "${SNITCH_LLVM_VERSION}" = "latest" && SNITCH_LLVM_VERSION=${latest_tag} || : ; \
+LLVM_TAR=riscv32-pulp-llvm-ubuntu2004-$(echo $SNITCH_LLVM_VERSION | cut -d '-' -f3-).tar.gz && \
+mkdir -p riscv-llvm && \
+echo "SNITCH_LLVM_VERSION=${SNITCH_LLVM_VERSION} LLVM_TAR=${LLVM_TAR} latest_tag=${latest_tag}" && \
+wget -qO- https://github.com/pulp-platform/llvm-project/releases/download/${SNITCH_LLVM_VERSION}/${LLVM_TAR} | \
+tar xvz --strip-components=1 -C riscv-llvm
+
+# 2. Stage: Filter python files to improve caching of the final container
+FROM ubuntu:${UBUNTU_VERSION} AS python_filter
+
+# Copy all files in `sw` and `util` and remove all non-python files
+COPY sw ./sw
+COPY util ./util
+
+RUN find sw -type f ! -name "*.py" -delete && \
+    find util -type f ! -name "*.py" -delete
```

Comment:
If I understand correctly, any change in `sw` or `util` would cause it to rebuild instead of fetch from the cache, right?

In this case, although there is no glob support in Docker, this [thread](https://stackoverflow.com/questions/35670907/docker-copy-with-file-globbing) specifies how one could use a `.dockerignore` file to achieve this purpose. We could maybe use this to make it more fine-grained.

What do you think about this? I would leave the decision up to you, we could even come back to this later on.

Edit:
Going through the file I now understand you are using this stage to implement solution 3 in the StackOverflow thread linked above. Maybe we can extend the comment a bit so we don't forget about this in the future.

Reply:
> If I understand correctly, any change in sw or util would cause it to rebuild instead of fetch from the cache, right?

Yes, exactly

> In this case, although there is no glob support in Docker, this [thread](https://stackoverflow.com/questions/35670907/docker-copy-with-file-globbing) specifies how one could use a .dockerignore file to achieve this purpose. We could maybe use this to make it more fine-grained.

Ah yes, `.dockerignore` might also be an option. I did not consider this, but is probably cleaner if it is possible.

> Edit:
Going through the file I now understand you are using this stage to implement solution 3 in the StackOverflow thread linked above. Maybe we can extend the comment a bit so we don't forget about this in the future.

Yes, that seems to be the same solution ChatGPT gave me😁

Reply:
A `.dockerignore` file would be much cleaner. I will implement this

Reply:
implemented in https://github.com/pulp-platform/snitch_cluster/pull/192/commits/08f4840678fc0f30530b5de56799c6e2158d8972